New course enrollment UX

singingwolfboy opened the pull request:
#### What are the relevant tickets?
Fixes #2467.

#### What's this PR do?
Changes the user experience for enrolling in courses through MicroMasters. Adds a message above the course listing in a program, explaining what the user's price is for these courses. Changes the buttons to say "Enroll Now" instead of displaying the course price. When the user clicks one of these "Enroll Now" buttons, the user sees a modal with two options: pay for the course, or audit it for free.

#### How should this be manually tested?
1. While enrolled in a program that has some courses that you are not yet enrolled in, visit the dashboard. Verify that the buttons to the right of the courses say "Enroll Now", and do not have any course price in the button text. Verify that there is no "Enroll and pay later" link. Click on the button, and verify that a dialog pops up with two choices: "Pay Now" and "Audit for Free & Pay Later". When clicked, the "Pay Now" button should redirect the user to the order summary page. When clicked, the "Audit" button should cause the user to be enrolled in the course on edX on the audit track.
2. Courses where you are already enrolled as an auditing user should instead have a button that says "Pay Now". Clicking on this button should not open a dialog, but should immediately redirect the user to the order summary page.
3. For a course without financial aid available, verify that the cost of the course is described in the blue box at the top of the course list. When the course enrollment dialog is opened, verify that the "Pay Now" button also includes the cost of the course.
4. Create a coupon for this course, and apply the coupon to your user account. Verify that the reduced price is correctly reflected in the text of the blue box above the course list, and in the "Pay Now" button.
5. For a course _with_ financial aid available, verify that the "Pay Now" button in the enrollment dialog is disabled before the user has calculated their financial aid. The text of the blue box should also direct the user to calculate their financial aid. Verify that the user can still audit the course.
6. After financial aid has been calculated, verify that the blue box and the "Pay Now" button should display the cost with financial aid applied. The "Pay Now" button should no longer be disabled.
7. For a course with both financial aid and a coupon, the price in the blue box and the "Pay Now" button should reflect both discounts.
8. Create a 100% off coupon for a program. Verify that clicking on the "Enroll Now" button on the dashboard for a course in that program does _not_ cause the enrollment dialog to appear, but instead redirects the user to the order summary page immediately.

#### Screenshots (if appropriate)
<img width="643" alt="course-price-message" src="https://cloud.githubusercontent.com/assets/132355/23184859/5abd23f8-f84f-11e6-9226-a5847bc43dc8.png">
<img width="585" alt="enroll dialog disabled" src="https://cloud.githubusercontent.com/assets/132355/23184875/66eec08c-f84f-11e6-9c25-6d44592bd280.png">
<img width="772" alt="enroll dialog enabled" src="https://cloud.githubusercontent.com/assets/132355/23184878/69e4a0cc-f84f-11e6-8d5a-f8e537e99be5.png">

![audit](https://cloud.githubusercontent.com/assets/132355/23221740/dd74c63a-f8f3-11e6-9a23-e65e3fee4131.png)

rhysyngsun commented on static/js/components/dashboard/CourseListCard.js:
```diff
@@ -11,54 +11,145 @@ import type {
+    if (coupon &&
+        coupon.content_type === COUPON_CONTENT_TYPE_PROGRAM &&
+        coupon.amount_type === COUPON_AMOUNT_TYPE_PERCENT_DISCOUNT &&
+        coupon.amount.equals(1)
```
Might be worth factoring out `isFree` from here: https://github.com/mitodl/micromasters/pull/2519/files#diff-e7226b0752d5be35f40e935a28554779R81 since we have this in two places now (albeit this additionally checks if it's a program coupon).

singingwolfboy replied:
Good thought. I've written a function called `isFreeCoupon` in `js/lib/coupon.js`, and I'm using that instead of these checks.

rhysyngsun commented:
@singingwolfboy could you add steps to test this manually?

singingwolfboy commented on static/js/components/dashboard/CourseListCard.js:
```diff
@@ -11,54 +11,145 @@ import type {
+    }
+
+    if (program.financial_aid_availability) {
+      const financialAid = program.financial_aid_user_info;
```
Done.

singingwolfboy commented on static/js/components/dashboard/CourseListCard.js:
```diff
@@ -11,54 +11,145 @@ import type {
+          </p>;
+        } else {
+          return <p className="price-message">
+            Your Personal Course Price is <strong>${ price.toString() } USD per course.</strong>
```
Oops, JSX isn't as smart as I'd like it to be about whitespace. The problem isn't the line ending with a period, it's the line ending with a starting or ending HTML tag. I think I've fixed it, but it might be worth double-checking.

singingwolfboy commented on static/js/components/dashboard/CourseListCard.js:
```diff
@@ -11,54 +11,145 @@ import type {
+
+    // no financial aid
+    if (coupon && coupon.content_type === COUPON_CONTENT_TYPE_PROGRAM) {
+      let isDiscount = (
```
Done.

singingwolfboy commented on static/js/components/dashboard/CourseListCard.js:
```diff
@@ -11,54 +11,145 @@ import type {
+        If you want to audit courses for FREE and upgrade later,
+        click Enroll Now then choose the audit option.
+      </p>;
+    }
```
Done.

roberthouse54 commented on static/js/components/CourseEnrollmentDialog.js:
```diff
@@ -0,0 +1,97 @@
+        </Button>
+      );
+    } else {
+      message = `You need to get a personalized course price before
```
Sorry, my mistake. Yes, make it capitalized like in the other mocks.

singingwolfboy commented:
@rhysyngsun: I think this is ready for another pass. I've also added screenshots and manual test instructions.

Please also note that I have not yet written any new automated tests for the functionality in this pull request. I've changed the existing automated tests to make sure they pass, including deleting automated tests for functionality that no longer exists as a result of this pull request. However, I still need to write new tests, which is probably why Codecov is failing my commits. I indicated that this PR was ready for review because it's been in progress for awhile now, and I wanted to get some feedback on it even before the tests were ready. I would welcome feedback about which tests to write, and how best to go about writing them. For example, do you think we need a Selenium test for this change?

rhysyngsun commented:
@singingwolfboy code looks good, I still have to manual test.

As far as the tests, the selenium tests don't contribute to JS test coverage because they're run via pytest. Even then, I think we're only using selenium as a smoke test at this point so I think traditional unit testing of the new code here is the best approach.

rhysyngsun commented on static/js/components/dashboard/CourseAction.js:
```diff
@@ -67,39 +69,39 @@ export default class CourseAction extends React.Component {
         {...buttonProps}
       >
-        {text}
+        Enroll Now
```
@singingwolfboy in my manual testing so far, this is now resulting in the learner seeing the "Enroll Now" button when they're already enrolled:

![selection_029](https://cloud.githubusercontent.com/assets/28598/23219774/f66d5874-f8ed-11e6-8c10-48fe97d2a1c8.png)

singingwolfboy replied:
Thanks for catching that! It should be fixed now. When viewing a course that you are already auditing, the button says "Pay Now", and clicking it redirects the user to the order summary page (rather than popping the dialog).

singingwolfboy commented:
I've added some new automated tests, and fixed the issues that @rhysyngsun found in his testing. This should be ready for another review pass.

rhysyngsun commented on static/js/components/dashboard/CourseListCard.js:
```diff
@@ -11,54 +11,179 @@ import type {
+        click Enroll Now then choose the audit option.
+      </p>;
+    }
+  }
```
I think this should have an explicit `return null;` at the end for FA applications not in the states above.

singingwolfboy replied:
It's not possible to reach the end of this function without hitting a return value. I've added Flow type annotations that this function _always_ returns a `React$Element<*>` type, to verify that. In pseudocode, the logic of this function is:
```
if (x) {
  if (y) {
    return React element
  } else {
    return React element
  }
} else {
  return React element
}
```

The code will always hit one of those three return values.

rhysyngsun commented:
Discussed in chat, but noting here for the record:

If I audit an FA course without a completed application I'm seeing that I'm allowed to pay for the courses anyway:

![selection_030](https://cloud.githubusercontent.com/assets/28598/23277002/57ced164-f9da-11e6-9a1e-689bf8302701.png)
